runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/tasks/AbstractDokkaTask.kt

```diff
@@ -203,14 +203,22 @@ abstract class AbstractDokkaTask : DefaultTask() {
     internal open fun generateDocumentation() {
         DokkaBootstrap(runtime, DokkaBootstrapImpl::class).apply {
             configure(buildDokkaConfiguration().toCompactJsonString(), createProxyLogger())
+            var throwableInGenerate: Throwable? = null
             /**
              * Run in a new thread to avoid memory leaks that are related to ThreadLocal (that keeps `URLCLassLoader`)
              * Currently, all `ThreadLocal`s leaking are in the compiler/IDE codebase.
              */
-            Thread { generate() }.apply {
+            Thread {
+                try {
+                    generate()
+                } catch (t: Throwable) {
+                    throwableInGenerate = t
+                }
+            }.apply {
                 start()
                 join()
             }
+            throwableInGenerate?.let { throw it }
```

Comment:
I'm not sure why Kotlin allows writing such code and what exactly happens "under the hood" when this is run on a JVM. Is it thread safe? I can't say for sure...

I do know that you cannot write such code in Java, it requires the used variable to be final or either effectively final for good reasons.

<details>
<summary>screenshots</summary>

![2023-04-06_23-11-53](https://user-images.githubusercontent.com/22685910/230494868-417c367c-442a-4921-83ba-dc4455a38555.png)
![2023-04-06_23-10-50](https://user-images.githubusercontent.com/22685910/230494873-1d9525ee-6be5-43fe-9e29-132692e28686.png)

</details>

To be on the safe side, I think we can use `AtomicReference` to hold the throwable, then it should be thread safe. 

While at it, we could use `UncaughtExceptionHandler` (which actually catches `Throwable`) instead of the try-catch, but it's optional :)3

To put the words into code:

```kotlin
val uncaughtExceptionHolder = AtomicReference<Throwable?>()
/**
 * Run in a new thread to avoid memory leaks that are related to ThreadLocal (that keeps `URLCLassLoader`)
 * Currently, all `ThreadLocal`s leaking are in the compiler/IDE codebase.
 */
Thread { generate() }.apply {
    setUncaughtExceptionHandler { _, throwable ->
        uncaughtExceptionHolder.set(throwable)
    }
    start()
    join()
}
uncaughtExceptionHolder.get()?.let { throw it }
```

Reply:
Thanks a lot for looking into it and providing such a detailed guidance.

> I'm not sure why Kotlin allows writing such code and what exactly happens "under the hood" when this is run on a JVM. Is it thread safe? I can't say for sure...

I took a look at the bytecode generated (see https://godbolt.org/z/6d47nPs4a) and under the hood is using [`ObjectRef`](https://github.com/JetBrains/kotlin/blob/b65a68ca0db564f4680fde19abea480425eaba0b/libraries/stdlib/jvm/runtime/kotlin/jvm/internal/Ref.java#L13), which is a `static final class` that holds an `element`.
That allows the use of `var throwableInGenerate` even though it's not really final.

Anyway, that doesn't guarantee thread safety. So your proposed solution using `AtomicReference` does indeed help with ensuring that the `throwable` is visible to all threads and it is set atomically.

Reply:
oh, it's nice to know what it's being compiled into, thanks for looking into it and sharing :)